README.md
# Computer Network Realtime streamming app
 


SourceCode/CHANGELOG.txt
v0 - v3: Implemented all requirement and extend 1,2,3
v4: 
	Implemented extend 4,5
	Add Change PlaybackSpeed
v5:
	Implemnted simple convert tool to make more videos for user to watch in the menu
	Fixed permanently disble "switch video" menu making all further SWITCH action failed, by removing the state check in function switchChannel() of Client.py
v6:
	Cleanup Code
	New bug: sudden drop in fps occasionally
v7:
	More cleanup

SourceCode/RtpPacket.py
#newly create file, author: Nguyen Quang Phuc
import time  


class RtpPacket:
    def decode(self, data):
        self.header = data[0:12]
        self.payload = data[12:]

        self.version = (self.header[0] & 0xC0) >> 6
        self.padding = (self.header[0] & 0x20) >> 5
        self.extension = (self.header[0] & 0x10) >> 4
        self.cc = self.header[0] & 0x0F

        self.marker = (self.header[1] & 0x80) >> 7
        self.pt = self.header[1] & 0x7F

        self.seqnum = int.from_bytes(self.header[2:4], 'big', signed = False)
        self.timestamps = int.from_bytes(self.header[4:8], 'big', signed = False)
        self.ssrc = int.from_bytes(self.header[8:12], 'big', signed = False)

        #self.frameLength = int.from_bytes(self.payload[0:5], 'big', signed = False)
        #self.frame = self.payload[5:self.frameLength + 5] #not sure, just in case payload has blank section at the end
        self.frame = self.payload

    def encode(self, version, padding, extension, cc, seqnum, marker, pt, ssrc, payload):
        self.payload = payload
        byte = version
        byte = byte << 1
        byte = byte | padding
        byte = byte << 1
        byte = byte | extension
        byte = byte << 4
        byte = byte | cc
        byte = byte << 1
        byte = byte | marker
        byte = byte << 7
        byte = byte | pt
        byte = byte << 16
        byte = byte | seqnum

        ts = int(time.time() * 1000)  #timestamp in milisecond
        ts_truncated = 0xFFFFFFFF & ts #only keep lower 32 bit 

        self.header = byte.to_bytes(4, 'big', signed=False) + ts_truncated.to_bytes(4, 'big', signed=False) + ssrc.to_bytes(4, 'big', signed=False)
    def getPacket(self):
        return self.header + self.payload
'''
version = 2
padding = 1
extension = 1
cc = 7
marker = 1
pt = 26 # MJPEG type
seqnum = 1
ssrc = 4 
		
rtpPacket = RtpPacket()
		
rtpPacket.encode(version, padding, extension, cc, seqnum, marker, pt, ssrc, b'\x00\x00\x00\x00\x03\xAB\x1C\x2A')

data = rtpPacket.getPacket()

print(data.hex(':'))

rtpPacket.decode(data)
print("Version: " + str(rtpPacket.version))
print("Padding: " + str(rtpPacket.padding))
print("Extension: " + str(rtpPacket.extension))
print("cc: " + str(rtpPacket.cc))
print("marker: " + str(rtpPacket.marker))
print("pt: " + str(rtpPacket.pt))
print("seqnum: " + str(rtpPacket.seqnum))
print("timestamps: " + str(rtpPacket.timestamps))
print("ssrc: " + str(rtpPacket.ssrc))
#print("FrameLenght: " + str(rtpPacket.frameLength))
print(rtpPacket.frame)
'''

SourceCode/VideoStream.py
import os

class VideoStream:
	def __init__(self, filename):
		self.filename = filename
		self.frameNum = 0
		self.indexTable = [0]
		try:
			self.file = open(filename, 'rb')
			self.buildIndex()
		except:
			raise IOError
	
	def buildIndex(self): #new - used to handle jump request more efficiently
		data = self.nextFrame()
		while data:
			self.indexTable.append(self.file.tell())
			data = self.nextFrame()
		self.file.seek(0, 0)
		self.frameNum = 0
		
	def nextFrame(self):
		"""Get next frame."""
		data = self.file.read(5) # Get the framelength from the first 5 bits
		if data: 
			#WOW. The first 5 byte is a string of digit...
			framelength = int(data)
							
			# Read the current frame
			data = self.file.read(framelength)
			self.frameNum += 1
		return data
		
	def frameNbr(self):
		"""Get frame number."""
		return self.frameNum #frameNum is actutally index in indexTable + 1
	
	def jumpTo(self, frameIndex):
		self.file.seek(self.indexTable[frameIndex], 0)
		self.frameNum = frameIndex

	def getFileSize(self):
		file_stats = os.stat(self.filename)
		return file_stats.st_size

	def getNumberOfFrame(self):
		return len(self.indexTable) 
		"""
		currentPos = self.file.tell() #save current position
		currentFrameNum = self.frameNum

		self.file.seek(0, 0) #move position to start to prepare for counting
		data = self.nextFrame()
		while data:
			data = self.nextFrame()
		self.file.seek(currentPos, 0) #move the stream position to current
		
		result = self.frameNum
		self.frameNum = currentFrameNum

		return result
		"""
	
	

SourceCode/ServerWorker.py
from random import randint
import sys, traceback, threading, socket

from VideoStream import VideoStream
from RtpPacket import RtpPacket

import time
import glob # to list all video files from the server's directory

class ServerWorker:
	SETUP = 'SETUP'
	PLAY = 'PLAY'
	PAUSE = 'PAUSE'
	TEARDOWN = 'TEARDOWN'
	DESCRIBE = 'DESCRIBE' #new
	JUMP = 'JUMP' #new
	SWITCH = 'SWITCH' #new
	CHANGESPEED = 'CHANGESPEED' #new

	INIT = 0
	READY = 1
	PLAYING = 2
	SWITCHING = 3 #new

	state = INIT

	OK_200 = 0
	FILE_NOT_FOUND_404 = 1
	CON_ERR_500 = 2
	
	clientInfo = {}
	
	def __init__(self, clientInfo):
		self.clientInfo = clientInfo
		self.clientInfo['frameDelay'] = 0.05
		
	def run(self):
		threading.Thread(target=self.recvRtspRequest).start()
	
	def recvRtspRequest(self):
		"""Receive RTSP request from the client."""
		connSocket = self.clientInfo['rtspSocket'][0]
		while True:            
			data = connSocket.recv(256)
			if data:
				print("Data received:\n" + data.decode("utf-8"))
				self.processRtspRequest(data.decode("utf-8"))
	
	def processRtspRequest(self, data):
		"""Process RTSP request sent from the client."""
		# Get the request type
		request = data.split('\n')
		line1 = request[0].split(' ')
		requestType = line1[0]
		
		# Get the media file name
		filename = line1[1]
		
		# Get the RTSP sequence number 
		seq = request[1].split(' ')		

		# Process SETUP request
		if requestType == self.SETUP:
			if self.state == self.INIT:
				# Update state
				print("processing SETUP\n")
				
				try:#modified: move 3 last lines into the trycatch
					self.clientInfo['videoStream'] = VideoStream(filename)
					self.state = self.READY

					# Generate a randomized RTSP session ID
					self.clientInfo['session'] = randint(100000, 999999)
				
					# Send RTSP reply
					#self.replyRtsp(self.OK_200, seq[1])
					reply = 'RTSP/1.0 200 OK\nCSeq: ' + seq[1] + '\nSession: ' + str(self.clientInfo['session']) + "\n"
					reply += "TotalFrame " + str(self.clientInfo['videoStream'].getNumberOfFrame())
					connSocket = self.clientInfo['rtspSocket'][0]
					connSocket.send(reply.encode())
				
					# Get the RTP/UDP port from the last line
					self.clientInfo['rtpPort'] = request[2].split(' ')[3]

				except IOError:
					self.replyRtsp(self.FILE_NOT_FOUND_404, seq[1])				
				
		
		# Process PLAY request 		
		elif requestType == self.PLAY:
			if self.state == self.READY or self.state == self.SWITCHING:
				print("processing PLAY\n")
				self.state = self.PLAYING
				
				# Create a new socket for RTP/UDP
				self.clientInfo["rtpSocket"] = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
				
				self.replyRtsp(self.OK_200, seq[1])
				
				# Create a new thread and start sending RTP packets
				self.clientInfo['event'] = threading.Event()
				self.clientInfo['worker']= threading.Thread(target=self.sendRtp) 
				self.clientInfo['worker'].start()
		
		# Process PAUSE request
		elif requestType == self.PAUSE:
			if self.state == self.PLAYING:
				print("processing PAUSE\n")
				self.state = self.READY
				
				self.clientInfo['event'].set()
			
				self.replyRtsp(self.OK_200, seq[1])
		
		# Process TEARDOWN request
		elif requestType == self.TEARDOWN:
			print("processing TEARDOWN\n")
			self.state = self.INIT #addition

			self.clientInfo['event'].set()
			
			self.replyRtsp(self.OK_200, seq[1])
			
			# Close the RTP socket
			self.clientInfo['rtpSocket'].close()
		
		# Process DESCRIBE request - extend 3
		elif requestType == self.DESCRIBE:
			print("processing DESCRIBE\n")
			try:
				sessionInfo = "v=2\n"
				sessionInfo += "o=" + self.clientInfo['rtspSocket'][0].getpeername()[0] + "\n"
				sessionInfo += "s=RTSP " + str(self.clientInfo['session']) + "\n"
				sessionInfo += "m=video " + str(filename) + " RTP/Mjpge\n"
				sessionInfo += "m=size " + str(self.clientInfo['videoStream'].getFileSize()) + " Bytes"

				reply = 'RTSP/1.0 200 OK\nCSeq: ' + seq[1] + '\nSession: ' + str(self.clientInfo['session']) + "\n"
				reply += "Content-Length: " + str(len(sessionInfo)) + "\n\n"
				reply += sessionInfo
				connSocket = self.clientInfo['rtspSocket'][0]
				connSocket.send(reply.encode())
			except IOError:
				self.replyRtsp(self.FILE_NOT_FOUND_404, seq[1])

		# Process JUMP request - extend 4
		elif requestType == self.JUMP:
			print("processing JUMP\n")
			try:
				self.replyRtsp(self.OK_200, seq[1])
				frameIndex = int(request[3].split(' ')[1])
				self.clientInfo['videoStream'].jumpTo(frameIndex)
			except IOError:
				self.replyRtsp(self.FILE_NOT_FOUND_404, seq[1])

		# Process SWITCH request - extend 5
		elif requestType == self.SWITCH:
			print("processing SWITCH\n")
			self.state == self.SWITCHING
			try:
				fileList = glob.glob('*.Mjpeg')

				reply = 'RTSP/1.0 200 OK\nCSeq: ' + seq[1] + '\nSession: ' + str(self.clientInfo['session']) + "\n"
				reply += "Videos:"
				for file in fileList:
					reply += " "+ file
				connSocket = self.clientInfo['rtspSocket'][0]
				connSocket.send(reply.encode())
			except IOError:
				self.replyRtsp(self.FILE_NOT_FOUND_404, seq[1])

		elif requestType == self.CHANGESPEED:
			print("processing CHANGESPEED\n")
			self.replyRtsp(self.OK_200, seq[1])
			newDelay = float(request[3].split(' ')[1])
			self.clientInfo['frameDelay'] = newDelay

	def sendRtp(self):
		"""Send RTP packets over UDP."""
		lastFrameTime = time.time()
		while True:
			currentFrameTime = time.time()
			if (currentFrameTime - lastFrameTime) > self.clientInfo['frameDelay']:
				lastFrameTime = currentFrameTime
				#self.clientInfo['event'].wait(0.05) I don't trust this clock
			
				# Stop sending if request is PAUSE or TEARDOWN
				if self.clientInfo['event'].isSet(): 
					break 
				
				data = self.clientInfo['videoStream'].nextFrame()
				if data: 
					frameNumber = self.clientInfo['videoStream'].frameNbr()
					try:
						address = self.clientInfo['rtspSocket'][1][0]
						port = int(self.clientInfo['rtpPort'])
						self.clientInfo['rtpSocket'].sendto(self.makeRtp(data, frameNumber),(address,port))
					except:
						print("Connection Error")
						#print('-'*60)
						#traceback.print_exc(file=sys.stdout)
						#print('-'*60)

	def makeRtp(self, payload, frameNbr):
		"""RTP-packetize the video data."""
		version = 2
		padding = 0
		extension = 0
		cc = 0
		marker = 0
		pt = 26 # MJPEG type
		seqnum = frameNbr
		ssrc = 0 
		
		rtpPacket = RtpPacket()
		
		rtpPacket.encode(version, padding, extension, cc, seqnum, marker, pt, ssrc, payload)
		
		return rtpPacket.getPacket()
		
	def replyRtsp(self, code, seq):
		"""Send RTSP reply to the client."""
		if code == self.OK_200:
			#print("200 OK")
			reply = 'RTSP/1.0 200 OK\nCSeq: ' + seq + '\nSession: ' + str(self.clientInfo['session'])
			connSocket = self.clientInfo['rtspSocket'][0]
			connSocket.send(reply.encode())
		
		# Error messages
		elif code == self.FILE_NOT_FOUND_404:
			#addition: inform the client about 404 error
			reply = 'RTSP/1.0 404 FILE_NOT_FOUND\nCSeq: ' + seq
			connSocket = self.clientInfo['rtspSocket'][0]
			connSocket.send(reply.encode())

			print("404 NOT FOUND")
		elif code == self.CON_ERR_500:
			print("500 CONNECTION ERROR")


SourceCode/ConvertTo5ByteMjpeg.py
#Author: Nguyen Quang Phuc - phuc.nguyenbk101@hcmut.edu.vn
#Just a simple tool to convert other video file to our special Mjpeg format
#Each jpeg is preceded by 5 byte 'string of digits' (that means the maximum size for a frame is 99,999 Byte)

print("Input filename: ")
f = input()

class VideoConvert:
	def __init__(self, filename):
		self.filename = filename

		self.file = open(filename, 'rb')
		self.data = self.file.read()

		self.modFile = open("modified_" + filename, 'wb')

		jpegStart = self.data.find(b'\xff\xd8') #start section of a jpeg
		jpegEnd = self.data.find(b'\xff\xd9') #end section of a jpeg
		print(str(jpegStart) + " - " + str(jpegEnd))
		while jpegStart != -1 and jpegEnd != -1:
			print(str(jpegStart) + " - " + str(jpegEnd))
			
			length = jpegEnd - jpegStart + 2
			self.modFile.write(str(length).encode('utf-8'))
			self.modFile.write(self.data[jpegStart:jpegEnd+2])

			self.data = self.data[jpegEnd+2:]
			jpegStart = self.data.find(b'\xff\xd8')#start section of a jpeg
			jpegEnd = self.data.find(b'\xff\xd9') #end section of a jpeg
VideoConvert(f)
input()


SourceCode/Client.py
#newly create file, author: Nguyen Quang Phuc

import tkinter as tk
from tkinter import messagebox as mb
from tkinter import ttk as ttk
import socket as sk
import threading 

from RtpPacket import RtpPacket #rtp decoding

from PIL import Image, ImageTk #jpeg displaying from IO stream
from io import BytesIO #jpeg loading from bytes to IOstream for display 

import time


class Client:
    INIT = 0
    READY = 1
    PLAYING = 2
    SWITCHING = 3

    state = INIT
    sequence = 0
    session = -1
    

    currentFrame = 0
    lossFrame = 0
    totalFrame = 0
    lastTime = 0
    maxFrameNum = 0
    
    def __init__(self, rootTK, serverAddr, serverPort, rtpPort, fileName):
        self.master = rootTK
        self.serverAddr = serverAddr
        self.serverPort = int(serverPort)
        self.rtpPort = int(rtpPort)
        self.fileName = fileName
        self.master.protocol("WM_DELETE_WINDOW", self.exit)
        #self.master.title("VIDEO SRC: " + fileName)

        #Setting up the GUI
        self.photo = ImageTk.PhotoImage(Image.open('Black.jpg'))

        self.videoFrame = tk.Frame(self.master, width = 600, height = 480)
        self.btnFrame = tk.Frame(self.master)
        self.statsLabel = tk.Label(self.master, text = "statistic", width = 30)

        self.controlScrollVar = tk.IntVar()
        self.videoScrollBar = tk.Scale(self.master, orient = "horizontal", showvalue = 0, sliderlength = 13, 
                                       width = 20, variable = self.controlScrollVar)
        #calling jump to only when user release the left mouse, to prevent firing more than once
        self.videoScrollBar.bind("<ButtonPress-1>", self.pauseBeforeJump) #pause before jump to avoid deadlock
        self.videoScrollBar.bind("<ButtonRelease-1>", self.jumpTo)

        self.statsLabel.pack(side = tk.RIGHT)
        self.videoFrame.pack()
        self.videoScrollBar.pack(fill = tk.X)
        self.btnFrame.pack()    

        self.imagePanel = tk.Label(master = self.videoFrame, image=self.photo)
        self.imagePanel.image = self.photo # keep a reference!
        self.imagePanel.pack() 

        self.playBtn = tk.Button(self.btnFrame, text = "Play", command = self.play)
        self.playBtn.grid(row = 0, column = 0)

        self.pauseBtn = tk.Button(self.btnFrame, text = "Pause", command = self.pause)
        self.pauseBtn.grid(row = 0, column = 1)

        self.teardownBtn = tk.Button(self.btnFrame, text = "STOP", command = self.teardown)
        self.teardownBtn.grid(row = 0, column = 2)

        self.describeBtn = tk.Button(self.btnFrame, text = "DESCRIBE", command = self.describe)
        self.describeBtn.grid(row = 0, column = 3)

        self.speedOptionMenuVar = tk.StringVar()
        self.speedOptionMenuVar.set("playback speed x1")
        self.speedOptionMenu = tk.OptionMenu(self.btnFrame, self.speedOptionMenuVar, "playback speed x0.1", 
            "playback speed x0.25", "playback speed x0.5", "playback speed x1", "playback speed x1.25", 
            "playback speed x1.5", "playback speed x1.75", "playback speed x2", "playback speed x2.5", 
            "playback speed x3")
        self.speedOptionMenu.grid(row = 1, column = 0, columnspan = 4)
        self.speedOptionMenuVar.trace("w", self.changeSpeed)

        #Extend 5 GUI
        self.videoListVar = tk.StringVar()
        self.videoListVar.set(self.fileName)
        self.videoList = tk.OptionMenu(self.btnFrame, self.videoListVar, self.fileName)
        self.videoList.bind('<ButtonRelease-1>', self.requestList) #called when the user click the dropdown menu
        self.videoListVarTraceID = self.videoListVar.trace("w", self.switchChannel)
        self.videoList.grid(row = 1, column = 4, columnspan = 1)     

        #Setting up socket RTPS/TCP
        self.rtpsSocket = sk.socket(sk.AF_INET, sk.SOCK_STREAM)
        self.rtpsSocket.connect((self.serverAddr, self.serverPort))          

    #input event handler
    def setup(self):
        if(self.state == self.INIT):
            #setup RTP/UDP socket
            self.rtpSocket = sk.socket(sk.AF_INET, sk.SOCK_DGRAM)
            self.rtpSocket.bind(('', self.rtpPort))

            self.sequence += 1

            message = "SETUP " + self.fileName + " RTPS/1.0\n"
            message += "CSeq: " + str(self.sequence) + "\n"
            message += "Transprot: RTP/UDP; client_port= " + str(self.rtpPort)
            self.rtpsSocket.send(message.encode())
        
            reply = self.rtpsSocket.recv(1024).decode("utf-8"); #1024 bytes buffer
            
            lines = reply.split('\n')
            if(lines[1] == "CSeq: " + str(self.sequence)):
                if(lines[0] == "RTSP/1.0 200 OK"):
                    self.state = self.READY
                    self.session = lines[2].split(' ')[1] 
                    self.maxFrameNum = int(lines[3].split(' ')[1])

                    self.videoScrollBar.config(to = self.maxFrameNum - 1)
                    self.controlScrollVar.set(0)

                    print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort)  + "\n" + reply)
                    print("----------------------\n")
                    #mb.showinfo("Reply" ,"From: " + str(self.serverAddr) + ":" + str(self.serverPort) 
                    #        + "\n" + reply               
                else:
                    mb.showerror("Error", lines[0])
            else:
                mb.showwarning("Warning", "Out of order sequence: " + lines[1] + ". Expecting " + str(self.sequence))
        else:
            mb.showwarning("Warning", "A connection was already setup!")

    def play(self):
        if(self.state == self.READY or self.state == self.SWITCHING):
            self.sequence += 1
            message = "PLAY " + self.fileName + " RTPS/1.0\n"
            message += "CSeq: " + str(self.sequence) + "\n"
            message += "Session: " + str(self.session)
            self.rtpsSocket.send(message.encode())
        
            reply = self.rtpsSocket.recv(1024).decode("utf-8");
            print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort) + "\n" + reply)
            print("----------------------\n")
            #mb.showinfo("Reply" ,"From: " + str(self.serverAddr) + ":" + str(self.serverPort) 
            #                + "\n" + reply)

            self.state = self.PLAYING
            self.rtpListener_PlayingFlag = True

            self.rtpSocket.settimeout(0.5)
            self.rtpListener = threading.Thread(target=self.receiveRtp)
            self.rtpListener.start()

        elif(self.state == self.INIT): #Extend 2
            self.setup()
            self.play()
        else:
            mb.showwarning("Warning", "The stream is already playing!")

    def pause(self):
        if(self.state == self.PLAYING):
            self.sequence += 1
            message = "PAUSE " + self.fileName + " RTPS/1.0\n"
            message += "CSeq: " + str(self.sequence) + "\n"
            message += "Session: " + str(self.session)
            self.rtpsSocket.send(message.encode())
        
            reply = self.rtpsSocket.recv(1024).decode("utf-8");
            print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort) + "\n" + reply)
            print("----------------------\n")
            #mb.showinfo("Reply" ,"From: " + str(self.serverAddr) + ":" + str(self.serverPort) 
            #                + "\n" + reply)

            self.rtpListener_PlayingFlag = False

            self.state = self.READY

        elif(self.state == self.INIT):
            mb.showwarning("Warning", "You need to setup the stream first!")
        else:
            mb.showwarning("Warning", "The stream is already paused!")

    def teardown(self): #Extend 3
        if(self.state == self.PLAYING or self.state == self.READY or self.state == self.SWITCHING):
            self.sequence += 1
            message = "TEARDOWN " + self.fileName + " RTPS/1.0\n"
            message += "CSeq: " + str(self.sequence) + "\n"
            message += "Session: " + str(self.session)
            self.rtpsSocket.send(message.encode())
        
            reply = self.rtpsSocket.recv(1024).decode("utf-8");
            print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort) + "\n" + reply)
            print("----------------------\n")
            #mb.showinfo("Reply" ,"From: " + str(self.serverAddr) + ":" + str(self.serverPort) 
            #                + "\n" + reply)

            self.rtpListener_PlayingFlag = False
            self.rtpSocket.close()

            self.currentFrame = 0
            self.lossFrame = 0
            self.totalFrame = 0

            self.state = self.INIT

        elif(self.state == self.INIT):
            mb.showwarning("Warning", "Nothing to teardown")
        #else:
           # mb.showwarning("Warning", "The stream cannot be teardown while pausing")

    def describe(self):
        if self.state == self.PLAYING or self.state == self.READY:
            self.sequence += 1
            message = "DESCRIBE " + self.fileName + " RTPS/1.0\n"
            message += "CSeq: " + str(self.sequence) + "\n"
            message += "Session: " + str(self.session)
            self.rtpsSocket.send(message.encode())
        
            reply = self.rtpsSocket.recv(1024).decode("utf-8");
            print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort) + "\n" + reply)
            print("----------------------\n")
            mb.showinfo("Reply" ,"From: " + str(self.serverAddr) + ":" + str(self.serverPort) + "\n" + reply)

    def jumpTo(self, e):#extend 4
        if self.state == self.READY:
            self.sequence += 1
            message = "JUMP " + self.fileName + " RTPS/1.0\n"
            message += "CSeq: " + str(self.sequence) + "\n"
            message += "Session: " + str(self.session) + "\n"
            message += "Frame: " + str(self.videoScrollBar.get())
            self.rtpsSocket.send(message.encode())
            #mb.showinfo("SEND" ,message)
            reply = self.rtpsSocket.recv(1024).decode("utf-8");
            print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort) + "\n" + reply)
            print("----------------------\n")

            self.currentFrame = self.videoScrollBar.get()
            self.play()

    def pauseBeforeJump(self, e):
        if self.state == self.PLAYING:
            self.pause()
    
    #Extend 5
    dropdownActived = False
    def requestList(self, e):
        if self.state == self.PLAYING:
            self.pause() #pause to avoid deadlock
        
        if self.state == self.READY and (not self.dropdownActived):
            self.sequence += 1
            message = "SWITCH " + self.fileName + " RTPS/1.0\n"
            message += "CSeq: " + str(self.sequence) + "\n"
            message += "Session: " + str(self.session)
            self.rtpsSocket.send(message.encode())
            #mb.showinfo("SEND" ,message)
            reply = self.rtpsSocket.recv(1024).decode("utf-8");

            lines = reply.split('\n')
            if(lines[1] == "CSeq: " + str(self.sequence)):
                if(lines[0] == "RTSP/1.0 200 OK"):
                    self.dropdownActived = True
                    self.state = self.SWITCHING

                    self.session = lines[2].split(' ')[1] 
                    fileList = lines[3].split(' ')[1:] #get list of files on server

                    self.videoList['menu'].delete(0, 'end')
                    for file in fileList:
                        tempLabel = file + " (current)" if file == self.fileName else file
                        self.videoList['menu'].add_command(label = tempLabel, command = (lambda x=file: self.videoListVar.set(x)))

                    print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort)  + "\n" + reply)
                    print("----------------------\n")

                    #mb.showinfo("Reply" ,"From: " + str(self.serverAddr) + ":" + str(self.serverPort) 
                    #        + "\n" + reply               
                else:
                    mb.showerror("Error", lines[0])
            else:
                mb.showwarning("Warning", "Out of order sequence: " + lines[1] + ". Expecting " + str(self.sequence))

    def switchChannel(self, *arg):
        self.dropdownActived = False
        if self.fileName == self.videoListVar.get():
            self.play()
            return
        #if file name is different than the current, set new file name so that future setup call will change to new channel
        self.fileName = self.videoListVar.get()
        self.teardown()

    #bonus feature - change playbackSpeed
    def changeSpeed(self, *arg):
        self.sequence += 1
        message = "CHANGESPEED " + self.fileName + " RTPS/1.0\n"
        message += "CSeq: " + str(self.sequence) + "\n"
        message += "Session: " + str(self.session) + "\n"
        speed = float(self.speedOptionMenuVar.get().split('x')[1])
        message += "Delay: " + str(0.05 / speed) #default delay between is 0.05

        self.rtpsSocket.send(message.encode())
        #mb.showinfo("SEND" ,message)
        reply = self.rtpsSocket.recv(1024).decode("utf-8");
        print("Reply From: " + str(self.serverAddr) + ":" + str(self.serverPort) + "\n" + reply)
        print("----------------------\n")

    def exit(self):
        self.teardown() #tell server to teardown the stream and stop sending
        self.rtpsSocket.close() #close the tcp connection
        self.master.destroy() #kill the application by stopping the mainloop


    #helper function
    def receiveRtp(self):
        maxFrameTime = 0.05 * self.maxFrameNum
        while self.rtpListener_PlayingFlag:
            try:
                data = self.rtpSocket.recv(1024*1024) #1MB buffer
                if data:
                    packet = RtpPacket()
                    packet.decode(data)

                    if(self.currentFrame + 1 != packet.seqnum):
                        self.lossFrame += 1
                    self.currentFrame = packet.seqnum
                    self.totalFrame += 1

                    current = time.time()
                    duration = current - self.lastTime
                    self.lastTime = current
                    speed = len(packet.frame) / duration
                    fps = 1/duration

                    currentFrameTime = 0.05 * self.currentFrame
                    remainingTime = maxFrameTime - currentFrameTime
                   
                    #Extend 1
                    stats = "RTP packet number: " + str(self.currentFrame) + "\n"
                    stats += "RTP loss packet: " + str(self.lossFrame) + "\n"
                    stats += "RTP packet loss rate: " + '{:.2f}'.format(self.lossFrame / self.totalFrame * 100) + "%\n"
                    stats += "Video data rate: " + '{:.2f}'.format(speed) + " Bytes/s\n"
                    stats += "Frame duration: " + '{:.0f}'.format(duration*1000) + " ms\n"
                    stats += "Frame per second: " + '{:.2f}'.format(fps) + "\n"
                    stats += "-----------------------\n"
                    #Extend 4
                    stats += "Remaining Time: " + '{:.0f}'.format(remainingTime) + " s\n"
                    stats += "Total Time: " + '{:.0f}'.format(maxFrameTime) + " s\n"
                    stats += "-----------------------\n"
                    stats += "VIDEO SRC: " + self.fileName + "\n"

                    self.statsLabel.configure(text = stats)

                    file_jpgdata = BytesIO(packet.frame) #PIL Image.open accept jpeg format from a file stream
                    self.photo = ImageTk.PhotoImage(Image.open(file_jpgdata))
                    self.imagePanel.configure(image = self.photo)
                    self.imagePanel.image = self.photo

                    #self.videoScrollBar.set(self.currentFrame - 1) #This will trigger the callback
                    self.controlScrollVar.set(self.currentFrame - 1) #This won't trigger the callback
            except:
                 pass
                 #print("Timeout Error!")




        
        